src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,139 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
                 properties: null,
                 SyntaxFacts.GetText(firstToken.Kind())));
         }
+
+        private static bool IsConsideredMultiLine(SyntaxNode statement, SyntaxNode embeddedStatement)
+        {
+            if (FormattingRangeHelper.AreTwoTokensOnSameLine(statement.GetFirstToken(), statement.GetLastToken()))
+            {
+                // The entire statement fits on one line. Examples:
+                //
+                //   if (something) return;
+                //
+                //   while (true) something();
+                return false;
+            }
+
+            if (!FormattingRangeHelper.AreTwoTokensOnSameLine(embeddedStatement.GetFirstToken(), embeddedStatement.GetLastToken()))
+            {
+                // The embedded statement does not fit on one line. Examples:
+                //
+                //   if (something)
+                //     obj.Method(   // <-- This embedded statement spans two lines.
+                //       arg);
+                return true;
+            }
+
+            var lastTokenBeforeEmbeddedStatement = embeddedStatement.GetFirstToken().GetPreviousToken();
+            if (!FormattingRangeHelper.AreTwoTokensOnSameLine(statement.GetFirstToken(), lastTokenBeforeEmbeddedStatement))
+            {
+                // The part of the statement preceding the embedded statement does not fit on one line. Examples:
+                //
+                //   for (int i = 0; // <-- The initializer/condition/increment are on separate lines
+                //        i < 10;
+                //        i++)
+                //     SomeMethod();
+                return true;
+            }
```

Comment:
this all makes sense.  i might just have a comment explaining that we think about these guys as two 'chunks'.  and that if, independently, each is on a single line, things are ok.

Reply:
(or maybe 3 chunks in the case of 'do while').

Reply:
➡️ New documentation added